iroh-dns-server/src/http/tls.rs

```diff
@@ -0,0 +1,173 @@
+            let certs = load_certs(cert_path)?;
+            let key = load_secret_key(key_path)?;
+            anyhow::Ok((certs, key))
+        })
+        .await??;
+
+        let config = config.with_single_cert(certs, secret_key)?;
+        let config = Arc::new(config);
+        // let acceptor = tokio_rustls::TlsAcceptor::from(config);
+        let acceptor = RustlsAcceptor::new(RustlsConfig::from_config(config));
+        Ok(Self::Manual(acceptor))
+    }
+
+    fn letsencrypt(
+        hostname: &str,
+        contact: &str,
+        is_production: bool,
+        dir: PathBuf,
+    ) -> Result<Self> {
+        let config = rustls::ServerConfig::builder()
+            .with_safe_defaults()
+            .with_no_client_auth();
+        let mut state = AcmeConfig::new(vec![hostname])
+            .contact([format!("mailto:{contact}")])
+            .cache_option(Some(DirCache::new(dir)))
+            .directory_lets_encrypt(is_production)
+            .state();
+
+        let config = config.with_cert_resolver(state.resolver());
+        let acceptor = state.acceptor();
+
+        tokio::spawn(
+            async move {
+                loop {
+                    match state.next().await.unwrap() {
+                        Ok(ok) => debug!("acme event: {:?}", ok),
+                        Err(err) => error!("error: {:?}", err),
+                    }
+                }
+            }
+            .instrument(info_span!("acme")),
+        );
+        let config = Arc::new(config);
+        let acceptor = AxumAcceptor::new(acceptor, config);
+        Ok(Self::LetsEncrypt(acceptor))
+    }
+}
+
+fn load_certs(filename: impl AsRef<Path>) -> Result<Vec<rustls::Certificate>> {
+    let certfile = std::fs::File::open(filename).context("cannot open certificate file")?;
+    let mut reader = std::io::BufReader::new(certfile);
+
+    let certs = rustls_pemfile::certs(&mut reader)?
+        .iter()
+        .map(|v| rustls::Certificate(v.clone()))
+        .collect();
+
+    Ok(certs)
+}
+
+fn load_secret_key(filename: impl AsRef<Path>) -> Result<rustls::PrivateKey> {
+    let keyfile = std::fs::File::open(filename.as_ref()).context("cannot open secret key file")?;
+    let mut reader = std::io::BufReader::new(keyfile);
+
+    loop {
+        match rustls_pemfile::read_one(&mut reader).context("cannot parse secret key .pem file")? {
+            Some(rustls_pemfile::Item::RSAKey(key)) => return Ok(rustls::PrivateKey(key)),
+            Some(rustls_pemfile::Item::PKCS8Key(key)) => return Ok(rustls::PrivateKey(key)),
+            Some(rustls_pemfile::Item::ECKey(key)) => return Ok(rustls::PrivateKey(key)),
+            None => break,
+            _ => {}
+        }
+    }
+
+    bail!(
+        "no keys found in {} (encrypted keys not supported)",
+        filename.as_ref().display()
+    );
+}
+
+fn escape_hostname(hostname: &str) -> Cow<'_, str> {
+    let unsafe_hostname_characters = regex::Regex::new(r"[^a-zA-Z0-9-\.]").unwrap();
```

Comment:
could be made static

Reply:
I think I took this code from iroh, so there as well likely
https://github.com/n0-computer/iroh/blob/319e9ccfda41af7f0b8a72337ef876875692eacd/iroh-net/src/bin/iroh-relay.rs#L131

Reply:
yep